### PR TITLE
Include DRS endpoint in integration test (#789, #790)

### PR DESCRIPTION
**Pull request by mjkrause**
* gets file ID using endpoint `/repository/files` with a filter and
query parameters to find small FAST-Q file

* GETs DRS URL, verifies status code is 200

* GETs URL to file from data object, verifies status code is 200

### CODE REVIEW

File: test/local_integration_test.py [hunk 1]
```diff
@@ -113,6 +120,15 @@ def check_bdbag_endpoint(self, lambda_endpoint: str, url: str):
...
                     logger.info(f'BDBag file {file_name} contains {num_rows} rows.')
 
+    def download_file_from_drs_response(self, response):
+        """Assert HTTP status code of file download is 200."""
+        json_data = response.json()['data_object']
+        url = first(json_data['urls'])['url']
+        file_name = json_data['name']
+        r = self.requests.get(url)
+        r.raise_for_status()
+        logger.info(f'Verified file {file_name} can be downloaded.')
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: I'd need some evidence that the file is actually a fastq.

File: test/local_integration_test.py [hunk 2]
```diff
@@ -214,3 +230,12 @@ def check_bundles_are_indexed(self, test_name: str, entity_type: str, indexed_bu
                             f'There was a problem during indexing an {entity_type} entity'
                             f' {hit["entryId"]}. Bundle(s) ({",".join(bundle_fqids)})'
                             f' have been indexed without the debug project name. Contains {project}')
+
+    def _get_file_uuid(self, lambda_endpoint: str, url: str):
+        """Returns file UUID of smallest fastq file in query."""
+        url = lambda_endpoint + url
+        response = self.requests.get(url)
+        response.raise_for_status()
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: And please stop duplicating functionality. There already is a function that does exactly those three lines.

**Comment by hannes-ucsc** [comment]
Commit title is not referencing both tickets.

File: test/local_integration_test.py [hunk 3]
```diff
@@ -214,3 +230,12 @@ def check_bundles_are_indexed(self, test_name: str, entity_type: str, indexed_bu
                             f'There was a problem during indexing an {entity_type} entity'
                             f' {hit["entryId"]}. Bundle(s) ({",".join(bundle_fqids)})'
                             f' have been indexed without the debug project name. Contains {project}')
+
+    def _get_file_uuid(self, lambda_endpoint: str, url: str):
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Changed and simplified URL handling, and added try/except to catch possible errors.

File: test/local_integration_test.py [hunk 4]
```diff
@@ -214,3 +230,12 @@ def check_bundles_are_indexed(self, test_name: str, entity_type: str, indexed_bu
...
+        """Returns file UUID of smallest fastq file in query."""
+        url = lambda_endpoint + url
+        response = self.requests.get(url)
+        response.raise_for_status()
+        hits = response.json()
+        assert hits['hits'] != [] and first(hits['hits'])['files'] != []
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Removed the assertion, rewrote def with try/except to catch and log errors.

File: test/local_integration_test.py (hunk 2 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Rewrote it. Now uses returned response from def `check_endpoint_is_working`.

File: test/local_integration_test.py (hunk 1 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Inserted two assertions, one for the file name in data object from the JSON response, and another on the file name of the file written to disk (to a temporary directory). IMHO, opening the file and checking its contents is beyond the scope of our responsibility.

Comment (mjkrause) [reply test/local_integration_test.py]: ^^^obsolete - now opening the files and checking characteristic bytes that indicate `.gz` and `.fastq`, respectively.

File: test/local_integration_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Removed try/except.

File: test/local_integration_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Removed try/except and log entries.

File: test/local_integration_test.py [hunk 5]
```diff
@@ -125,6 +137,26 @@ def check_bdbag(self, response: bytes):
                     self.assertTrue(num_rows > 1)
                     logger.info(f'BDBag file {file_name} contains {num_rows} rows.')
 
+    def download_file_from_drs_response(self, response_content):
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: Please make the signature consistent with the existing methods that consume a response.

Just in general, I find it annoying when contributors partially inject their preferences into the code base. Either change the signature for all methods of the same kind—and be ready to defend your decision—or stick with what's already there. You're not making it easier for anyone by having things your way only in the parts you change. Above all, we value consistency with precedent.

File: test/local_integration_test.py [hunk 6]
```diff
@@ -125,6 +137,26 @@ def check_bdbag(self, response: bytes):
...
+        json_data = json.loads(response_content)['data_object']
+        file_url = first(json_data['urls'])['url']
+        file_name = json_data['name']
+        self.assertTrue('fastq' in file_name.lower())
+        r_content = self.check_endpoint_is_working(file_url, '')
+        # The FASTQ file should be gzip compressed, check first two signature bytes.
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: True, it's unnecessary, and potentially risky if it is a FASTQ but `fastq` isn't in the name. Opening the file to look for signatures supersedes the outcome of this assertion anyways. Removed.

[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: I moved up the log message so it executes before the assertions on the gz file type (as suggested below).

File: test/local_integration_test.py [hunk 7]
```diff
@@ -125,6 +137,26 @@ def check_bdbag(self, response: bytes):
...
+        self.assertTrue(r_content[0:1].hex() == '1f')  # first byte
+        self.assertTrue(r_content[1:2].hex() == '8b')  # second byte
+        logger.info(f'Verified file type of {file_name} as gzip.')
+        # Check signature of FASTQ file.
+        with gzip.open(BytesIO(r_content)) as buf:
+            fastq = buf.read()
+        lines = fastq.splitlines()
+        # Assert first character of first and third lines of file (see https://en.wikipedia.org/wiki/FASTQ_format).
+        self.assertTrue(lines[0].decode().startswith('@'))
+        self.assertTrue(lines[2].decode() == '+')
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: I changed that assertion to `...startswith(b'@')`, consistent with the format description.

File: test/local_integration_test.py [hunk 8]
```diff
@@ -125,6 +137,26 @@ def check_bdbag(self, response: bytes):
...
+        logger.info(f'Unzipped file {file_name} and verified it to be a FASTQ file.')
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Moved the log message above the assertions.

File: test/local_integration_test.py [hunk 9]
```diff
@@ -226,3 +258,9 @@ def check_bundles_are_indexed(self, test_name: str, entity_type: str, indexed_bu
                             f'There was a problem during indexing an {entity_type} entity'
                             f' {hit["entryId"]}. Bundle(s) ({",".join(bundle_fqids)})'
                             f' have been indexed without the debug project name. Contains {project}')
+
+    @classmethod
+    def _get_file_uuid(cls, response_content):
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Done.

File: test/local_integration_test.py (hunk 5 above)
Comment (mjkrause) [reply test/local_integration_test.py]: I didn't intent to inject anything and I, too, value consistency and will try to implement it going forward. I changed it in all three _def_s that consume the response content to `(..., response_content: bytes, ...)`. That makes it clear that the argument is only the response content (similar to variable name `response_json`, used in some other _def_ in that module).

File: test/local_integration_test.py [hunk 10]
```diff
@@ -125,6 +137,26 @@ def check_bdbag(self, response: bytes):
...
+        # The FASTQ file should be gzip compressed, check first two signature bytes.
+        self.assertTrue(r_content[0:1].hex() == '1f')  # first byte
+        self.assertTrue(r_content[1:2].hex() == '8b')  # second byte
+        logger.info(f'Verified file type of {file_name} as gzip.')
+        # Check signature of FASTQ file.
+        with gzip.open(BytesIO(r_content)) as buf:
+            fastq = buf.read()
+        lines = fastq.splitlines()
+        # Assert first character of first and third lines of file (see https://en.wikipedia.org/wiki/FASTQ_format).
+        self.assertTrue(lines[0].decode().startswith('@'))
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Done.

File: test/local_integration_test.py [hunk 11]
```diff
@@ -86,6 +89,15 @@ def test_webservice_and_indexer(self):
                     response = self.check_endpoint_is_working(config.service_endpoint(), path, query)
                     validator(response)
 
+        filters = {"file":{"organ":{"is":["Brain"]},"fileFormat":{"is":["fastq"]}}}
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]:
```suggestion
        filters = {"file":{"organ":{"is":["brain"]},"fileFormat":{"is":["fastq"]}}}
```

File: test/local_integration_test.py (hunk 5 above)
Comment (hannes-ucsc) [reply test/local_integration_test.py]: You forgot the type annotation. And please use the name `response`, not `response_content`.

File: test/local_integration_test.py (hunk 11 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Good catch - changed. Thanks.

Comment (mjkrause) [reply test/local_integration_test.py]: After changing organ to `brain` I needed to change `fastq` to `fastq.gz` in addition. If organ is `Brain`, both file formats are available, for `brain` only `fastq.gz`, not `fastq` is in the index.

File: test/local_integration_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply test/local_integration_test.py]: Please address.

[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply test/local_integration_test.py]: Please address.

File: test/local_integration_test.py [hunk 12]
```diff
@@ -125,6 +137,20 @@ def check_bdbag(self, response: bytes):
                     self.assertTrue(num_rows > 1)
                     logger.info(f'BDBag file {file_name} contains {num_rows} rows.')
 
+    def download_file_from_drs_response(self, response: bytes):
+        json_data = json.loads(response)['data_object']
+        file_url = first(json_data['urls'])['url']
+        file_name = json_data['name']
+        response = self.check_endpoint_is_working(file_url, '')
+        # Check signature of FASTQ file.
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: This comment doesn't really apply. We don't just check the signature of the FASTQ. We open it.

**Pull request merged**